editor/src/clj/dynamo/graph.clj

```diff
@@ -1156,8 +1172,13 @@
                                         (map #(vector % serial-id)
                                              (override-originals basis original-id))))
                               replacements)
-         include-arc?   (partial ig/arc-endpoints-p (partial contains? serial-ids))
-         serialize-arc  (partial serialize-arc serial-ids)
+         include-arc?   (partial ig/arc-endpoints-p
+                                 (fn [id label]
+                                   (or (contains? serial-ids id)
+                                       (and (contains? external-refs id)
+                                            (contains? (get external-labels id) label)))))
```

Comment:
This is a bit weird. The `fn` is called with both the `[source-id source-label]` and `[target-id target-label]`, and must return true for both ends of the `Arc` in order to be included. That means the `external-labels` must include both the source-labels and target-labels of both ends like this:
```
#{source-label-in-project
  target-label-in-consumer
  source-label-in-consumer
  target-label-in-project}
```
It may be the case that the outputs on the project node match the input name of the consumer, but that seems like a strange limitation. And it also opens up the possibility of including unwanted `Arcs`.

I really think the interface for externals should only concern itself with outputs or inputs on the project node in our example, and it shouldn't matter what the consumer inputs or outputs are named. I don't think we should be using `ig/arc-endpoints-p` at all here, but look at the `Arcs` directly. And maybe the interface should be something like this to distinguish between inputs and outputs? (I'm merging `:external-refs` and `:external-labels` here)
```
:external-refs {project {:serial-id :project
                         :inputs #{:input-on-project-node}
                         :outputs #{:output-on-project-node}}}
```

Reply:
Is it possible you are misunderstanding how it works? We don't care how copied node's outputs and inputs are named, and labels only concern the project node. `ig/arc-endpoints-p` is true only when the pred is true for both ends of the arc. This means endpoints of an arc should either:
- belong to a node we are copying (`(contains? serial-ids id)`);
- be a project node (`(contains? external-refs id)`) with the project's label (`(contains? (get external-labels id) label)`).

There is no expectation anywhere for copied node's label to match the project's label, and we don't specify copied node's labels.

Reply:
Ah, I see now. We won't even get there if it is a node that was part of the subtree. 👍